File: 31-next-permutation/31-next-permutation.java
class Solution {
    public void nextPermutation(int[] nums) {
        
        int pivot1 = -1;
        for(int i = nums.length - 2; i >= 0; i--){
            if(nums[i] < nums[i + 1]){
                pivot1 = i;
                break;
            }
        }
        
        if(pivot1 == -1){
            reverse(0, nums);
            return;
        }
        for(int j = nums.length - 1; j > pivot1; j--){
            if(nums[j] > nums[pivot1]){
                swap(j, pivot1, nums);
                break;
            }
        }
        
        reverse(pivot1 + 1, nums);
        
        
    }
    
    public static void swap(int i, int j, int[] nums){
            int tmp = nums[i];
            nums[i] = nums[j];
            nums[j] = tmp;
    }
    
    public void reverse(int i, int[] nums){
        int j = nums.length - 1;
        while(i < j){
            swap(i, j, nums);
            i++;
            j--;
        }
        
    }
}

File: 31-next-permutation/31-next-permutation.py
class Solution:
    def nextPermutation(self, nums: List[int]) -> None:
        """
        Do not return anything, modify nums in-place instead.
        """
        
        i = len(nums) - 1
        
        while i - 1 >= 0 and nums[i - 1] >= nums[i]: 
            i -= 1
        
        if i == 0:
            nums.reverse()
            return
        
        #find the decreaseing sequence so its biggest ending, we have to change the previous one
        j = i
        
        
        
        #we want to reverse the ending to get the smalles tail. But we also want the swap to make the smallest tail.
        #so we want the fronter digits to be done (in reverse order). but in the original order, the rightmost larger num should be swapped
        while j + 1 < len(nums) and nums[j + 1] > nums[i - 1]:
            j += 1
        
        # j is the rightmost larger number, swap
        
        nums[j], nums[i - 1] = nums[i - 1], nums[j]
        nums[i:] = nums[i:][::-1]
        
        
        
        
        